src/v/v8_engine/executor.cc

```diff
@@ -0,0 +1,169 @@
+/*
+ * Copyright 2021 Vectorized, Inc.
+ *
+ * Licensed as a Redpanda Enterprise file under the Redpanda Community
+ * License (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * https://github.com/vectorizedio/redpanda/blob/master/licenses/rcl.md
+ */
+
+#include "v8_engine/executor.h"
+
+#include "seastarx.h"
+
+#include <seastar/core/alien.hh>
+#include <seastar/core/do_with.hh>
+#include <seastar/core/future.hh>
+#include <seastar/core/gate.hh>
+#include <seastar/core/semaphore.hh>
+#include <seastar/core/sleep.hh>
+#include <seastar/core/smp.hh>
+#include <seastar/util/later.hh>
+
+#include <chrono>
+#include <memory>
+
+namespace v8_engine {
+
+namespace internal {
+
+// condition
+
+condition::condition() noexcept
+  : _file_desc{ss::file_desc::eventfd(0, 0)}
+  , _fd(_file_desc.get())
+  , _pollable_fd{std::move(_file_desc)} {}
+
+ss::future<> condition::wait() {
+    return _pollable_fd
+      .read_some(reinterpret_cast<char*>(&_event), sizeof(_event))
+      .discard_result();
+}
+
+void condition::notify() { ::eventfd_write(_fd, 1); }
+
+// spsc_queue
+
+spsc_queue::spsc_queue()
+  : _ss_mutex(ss::smp::count)
+  , _items(ss::smp::count)
+  , _ss_mutex_shard(ss::this_shard_id()) {}
+
+void spsc_queue::close() {
+    is_stopped = true;
+    _ss_mutex.broken(ss::gate_closed_exception());
+    _has_element_cv.notify_all();
+}
+
+ss::future<> spsc_queue::push(work_item* item) {
+    co_await _ss_mutex.wait(1);
```

Comment:
I am wondering if we can use `condition` in here instead of a mutex. it would be something like
```c++
while(full()){
 co_await _cond;
}
_items.push()

```

Reply:
Do we need to add exponential backoff?

Reply:
no I think we do not